### PR TITLE
add initial profile file provider support

### PR DESCRIPTION
**Pull request by rcoh**
## Motivation and Context
Currently, only the parsing side of profile file providers is implemented. This adds the execution side of profile file providers. Currently, we also have a number of manually generated test cases that track HTTP behavior. This introduces DVR, which is a higher quality abstraction for HTTP verification testing.
## Description

This commit establishes baseline support for profile file providers. The only named credential source currently supported is `Environment` but we'll be adding more sources in the future. To support testing, the `dvr` module was added to smithy-client. There are some limitations in this which will be addressed going forward, primarily around time control so we can do request verification.

I cleaned up the DVR work quite a bit, I suspect we'll do a few more rounds on that so feel free to skim at the moment.

## Testing
1. Unit tests for specific error cases
2. DVR tests for e2e behavior verification
3. Test coverage was verified.

## Checklist
- [x] I have updated the CHANGELOG

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: rust-runtime/smithy-client/src/dvr/replay.rs [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+use std::collections::{HashMap, VecDeque};
+use std::error::Error;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::sync::{Arc, Mutex};
+use std::task::{Context, Poll};
+
+/// Replay traffic recorded by a [`RecordingConnection`](RecordingConnection)
+#[derive(Clone, Debug)]
+pub struct ReplayingConnection {
+    events: Arc<Mutex<HashMap<ConnectionId, VecDeque<Event>>>>,
```
Comment (jdisanti) [inline rust-runtime/smithy-client/src/dvr/replay.rs:20]: Do we need a map of connection id -> events, or can this just be a vec of a vec?

File: aws/rust-runtime/aws-auth-providers/src/profile.rs [hunk 2]
```diff
@@ -22,20 +22,72 @@
...
+            CredentialsError::Unhandled(format!("failed to load: {}", &err).into())
+        })?;
+        let mut creds = match inner
+            .base()
+            .provide_credentials()
+            .instrument(tracing::info_span!("load_base_credentials"))
+            .await
+        {
+            Ok(creds) => {
+                tracing::info!(creds = ?creds, "loaded base credentials");
```
[the remark this answers is not in the document]
Comment (rcoh) [reply aws/rust-runtime/aws-auth-providers/src/profile.rs:68]: correct, credentials have a safe debug impl

File: aws/rust-runtime/aws-auth-providers/src/profile.rs [hunk 3]
```diff
@@ -22,20 +22,72 @@
...
+                creds
+            }
+            Err(e) => {
+                tracing::warn!(error = %e, "failed to load base credentials");
+                return Err(e);
```
[the remark this answers is not in the document]
Comment (rcoh) [reply aws/rust-runtime/aws-auth-providers/src/profile.rs:73]: I don't think so, currently there isn't any logging on the top level credential provider

File: aws/rust-runtime/aws-auth-providers/src/profile.rs [hunk 4]
```diff
@@ -22,20 +22,72 @@
...
+        for provider in inner.chain().iter() {
+            let next_creds = provider
+                .credentials(creds, &self.client_config)
+                .instrument(tracing::info_span!("load_assume_role", provider = ?provider))
+                .await;
+            match next_creds {
+                Ok(next_creds) => creds = next_creds,
+                Err(e) => {
+                    tracing::warn!(provider = ?provider, "failed to load assume role credentials");
+                    return Err(e);
```
[the remark this answers is not in the document]
Comment (rcoh) [reply aws/rust-runtime/aws-auth-providers/src/profile.rs:85]: I don't believe so, there isn't top level logging of the failure. (And if there was, I think it's still valuable to have this site specific logging of the problem)

File: rust-runtime/smithy-client/src/dvr/replay.rs (hunk 1 above)
Comment (rcoh) [reply rust-runtime/smithy-client/src/dvr/replay.rs:20]: discussed offline, you could do it with Vec<Vec<..>> but it then becomes more annoying to reassemble the requests afterword

File: rust-runtime/smithy-client/src/dvr/record.rs [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+    pub(crate) num_events: Arc<AtomicUsize>,
+    pub(crate) inner: S,
+}
+
+impl RecordingConnection<hyper::Client<HttpsConnector<HttpConnector>, SdkBody>> {
+    /// Construct a recording connection wrapping a default HTTPS implementation
+    #[cfg(feature = "hyper-rustls")]
+    pub fn https() -> Self {
+        let https = hyper_rustls::HttpsConnector::with_native_roots();
+        let client = hyper::Client::builder().build::<_, SdkBody>(https);
```
[the remark this answers is not in the document]
Comment (rcoh) [reply rust-runtime/smithy-client/src/dvr/record.rs:39]: or us `::new()` and BYO connector

**Review by jdisanti — APPROVED** [review]

**Pull request merged**